src/core/ext/filters/client_channel/lb_policy/weighted_round_robin/weighted_round_robin.cc

```diff
@@ -0,0 +1,955 @@
+      WeightedRoundRobin* p = static_cast<WeightedRoundRobin*>(policy());
+      p->Unref(DEBUG_LOCATION, "subchannel_list");
+    }
+
+    // Updates the counters of subchannels in each state when a
+    // subchannel transitions from old_state to new_state.
+    void UpdateStateCountersLocked(
+        absl::optional<grpc_connectivity_state> old_state,
+        grpc_connectivity_state new_state);
+
+    // Ensures that the right subchannel list is used and then updates
+    // the aggregated connectivity state based on the subchannel list's
+    // state counters.
+    void MaybeUpdateAggregatedConnectivityStateLocked(
+        absl::Status status_for_tf);
+
+   private:
+    std::string CountersString() const {
+      return absl::StrCat("num_subchannels=", num_subchannels(),
+                          " num_ready=", num_ready_,
+                          " num_connecting=", num_connecting_,
+                          " num_transient_failure=", num_transient_failure_);
+    }
+
+    size_t num_ready_ = 0;
+    size_t num_connecting_ = 0;
+    size_t num_transient_failure_ = 0;
+
+    absl::Status last_failure_;
+  };
+
+  // A picker that performs WRR picks with weights based on
+  // endpoint-reported utilization and QPS.
+  class Picker : public SubchannelPicker {
+   public:
+    Picker(RefCountedPtr<WeightedRoundRobin> wrr,
+           WeightedRoundRobinSubchannelList* subchannel_list);
+
+    ~Picker() override;
+
+    PickResult Pick(PickArgs args) override;
+
+    void Orphan() override;
+
+   private:
+    // A call tracker that collects per-call endpoint utilization reports.
+    class SubchannelCallTracker : public SubchannelCallTrackerInterface {
+     public:
+      explicit SubchannelCallTracker(RefCountedPtr<AddressWeight> weight)
+          : weight_(std::move(weight)) {}
+
+      void Start() override {}
+
+      void Finish(FinishArgs args) override;
+
+     private:
+      RefCountedPtr<AddressWeight> weight_;
+    };
+
+    // Info stored about each subchannel.
+    struct SubchannelInfo {
+      SubchannelInfo(RefCountedPtr<SubchannelInterface> subchannel,
+                     RefCountedPtr<AddressWeight> weight)
+          : subchannel(std::move(subchannel)), weight(std::move(weight)) {}
+
+      RefCountedPtr<SubchannelInterface> subchannel;
+      RefCountedPtr<AddressWeight> weight;
+    };
+
+    // Returns the index into subchannels_ to be picked.
+    size_t PickIndex();
+
+    // Builds a new scheduler and swaps it into place, then starts a
+    // timer for the next update.
+    void BuildSchedulerAndStartTimerLocked()
+        ABSL_EXCLUSIVE_LOCKS_REQUIRED(&mu_);
+
+    RefCountedPtr<WeightedRoundRobin> wrr_;
+    const bool use_per_rpc_utilization_;
+    const Duration weight_update_period_;
+    const Duration weight_expiration_period_;
+    const Duration blackout_period_;
+    std::vector<SubchannelInfo> subchannels_;
+
+    Mutex mu_;
+    uint32_t scheduler_state_ ABSL_GUARDED_BY(&mu_);
```

Comment:
By persisent I meant for the lifetime of the WeightRoundRobinPolicy as in Stubby. I misread that the counter is per-scheduler in the previous comment but I think it does not change the point I wanted to make that keeping next_seq_ stable the whole time always benefits. This section also mentions this is an important feature: go/static-stride-scheduler#concurrency.

For example, since we use a map to dedup addresses the order is stable as long as the subchannel set is the same, and keeping next_seq_ persisent across pickers clearly helps improving the stability. This includes resolver updates with the same server address or a flapping subchannel (not perfect). In all I don't see a reason not to do what Stubby does here.

class WRR { 
   atomic<uint32_t> next_seq_ { rand() };
   uint32_t NextSeq() { return next_seq_ .fetch_add(1); }
   void UpdateWeights(..) {
       StaticStrideScheduler::Make(.., [this](){ return NextSeq(); });
   }
}

Reply:
I think we already discussed this extensively in the design.  Whenever we update the picker, the list of subchannels can change (e.g., because we got a new address list from the channel, or because a subchannel transitioned into or out of READY state), so the indexes of the subchannels change.  When the indexes change, there is no benefit to retaining the scheduler state, because it may no longer point to the same place in the rotation that it had previously pointed to.

Note that this is exactly consistent with how RR works in gRPC: whenever we generate a new picker, we [start from a random index in the list](https://github.com/grpc/grpc/blob/7fd808f6f9d12e3e63dc299984567aae591cf2eb/src/core/ext/filters/client_channel/lb_policy/round_robin/round_robin.cc#L212).  I think it makes sense to do the same thing in WRR until and unless it proves to be a problem.  Let's not add complexity that we don't know if we're actually going to need.

(To be fair, there are some cases today where we update the picker without the subchannel indexes actually changing.  This can happen whenever a subchannel changes from one non-READY state to another, such as going from IDLE to CONNECTING.  In these cases, we do wind up needlessly losing the scheduler state, and we could optimize that by avoiding a picker update in those cases.  But again, I don't think we should add the extra complexity here until and unless we know we're going to need it.)

Reply:
We also discussed the fact that the scheduler treats a weight of 0 as "average weight", not as "do not pick this address", which means that we have to change the indexes whenever a subchannel transitions into or out of state READY.

Reply:
As per our conversation, I've changed this to store the scheduler state in the LB policy itself so that it can be retained across pickers.

I'm still not sure this will actually yield the same benefit for gRPC that it does for Stubby, because gRPC will not order the addresses the same way that Stubby does, especially when backend tasks move around.  But it doesn't add too much complexity, and it does pave the way for #31973.